params/networkhelper/provider_utils.go

```diff
@@ -0,0 +1,184 @@
+package networkhelper
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/status-im/status-go/params"
+)
+
+// MergeProvidersPreservingUsersAndEnabledState merges new embedded providers with the current ones,
+// preserving user-defined providers and maintaining the Enabled state.
+func MergeProvidersPreservingUsersAndEnabledState(currentProviders, newProviders []params.RpcProvider) []params.RpcProvider {
+	// Create a map for quick lookup of the Enabled state by Name
+	enabledState := make(map[string]bool)
+	for _, provider := range currentProviders {
+		enabledState[provider.Name] = provider.Enabled
+	}
+
+	// Update the Enabled field in newProviders if the Name matches
+	for i := range newProviders {
+		if enabled, exists := enabledState[newProviders[i].Name]; exists {
+			newProviders[i].Enabled = enabled
+		}
+	}
+
+	// Retain current providers of type UserProviderType and add them to the beginning of the list
+	var mergedProviders []params.RpcProvider
+	for _, provider := range currentProviders {
+		if provider.Type == params.UserProviderType {
+			mergedProviders = append(mergedProviders, provider)
+		}
+	}
+
+	// Add the updated newProviders
+	mergedProviders = append(mergedProviders, newProviders...)
+
+	return mergedProviders
+}
+
+// ToggleUserProviders enables or disables all user-defined providers and disables other types.
+func ToggleUserProviders(providers []params.RpcProvider, enabled bool) []params.RpcProvider {
+	for i := range providers {
+		if providers[i].Type == params.UserProviderType {
+			providers[i].Enabled = enabled
+		} else {
+			providers[i].Enabled = !enabled
+		}
+	}
+	return providers
+}
+
+// GetEmbeddedProviders returns the embedded providers from the list.
+func GetEmbeddedProviders(providers []params.RpcProvider) []params.RpcProvider {
+	var embeddedProviders []params.RpcProvider
```

Comment:
and please check other places, there're more in this PR
```suggestion
    embeddedProviders := make([]params.RpcProvider, len(providers))
```

Reply:
fixed